csrc/multidevice/lower_communication.cpp

```diff
@@ -0,0 +1,320 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+#include <ir/interface_nodes.h>
+#include <multidevice/device_mesh.h>
+#include <multidevice/lower_communication.h>
+#include <multidevice/pipeline.h>
+
+namespace nvfuser {
+
+namespace {
+
+// Returns whether a TensorView has its first axis parallelized on Didx
+// Checks that the other axis are not parallelized on Didx
+bool isParallelD(TensorView* tv) {
+  std::vector<bool> is_parallel_d;
+  for (IterDomain* id : tv->getRootDomain()) {
```

Comment:
Should this be `getLeafDomain`?

Reply:
I honestly don't know what makes the first dim special than other dims.

Reply:
> Should this be `getLeafDomain`?

If you say so, it is not so clear to me to be honest, so I trust your call

> I honestly don't know what makes the first dim special than other dims.

I restrict the support of device parallelization to the first dim for now just to move in smaller steps. I am sure it avoids some complication, while being still general enough to support interesting cases.
My plan is to add parallelization of other dims later

Reply:
Maybe for now
```
NVF_ERROR(tv->getMaybeRFactorDomain() == tv->getLeafDomain());
```

Reply:
ok ! In addition to this check, should we replace `getRootDomain` with `getLeafDomain` as you suggested above ?

Reply:
Yes